cloud/openstack/os_port.py

```diff
@@ -0,0 +1,357 @@
+                       for match in module.params['fixed_ips']):
+                return True
+
+    return False
+
+
+def _system_state_change(module, port):
+    state = module.params['state']
+    if state == 'present':
+        if not port:
+            return True
+        return _needs_update(module, port)
+    if state == 'absent' and port:
+        return True
+    return False
+
+
+def _compose_port_args(module):
+    port_kwargs = {}
+    optional_parameters = ['name',
+                           'fixed_ips',
+                           'admin_state_up',
+                           'mac_address',
+                           'allowed_address_pairs',
+                           'extra_dhcp_opt',
+                           'device_owner',
+                           'device_id']
+    for optional_param in optional_parameters:
+        if module.params[optional_param] is not None:
+            port_kwargs[optional_param] = module.params[optional_param]
+
+    if module.params['no_security_groups']:
+        port_kwargs['security_groups'] = []
+    elif module.params['security_groups']:
+        port_kwargs['security_groups'] = module.params[
+            'security_groups'].split(',')
+
+    return port_kwargs
+
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        network=dict(required=False),
+        name=dict(required=False),
+        fixed_ips=dict(default=None),
+        admin_state_up=dict(default=None),
+        mac_address=dict(default=None),
+        security_groups=dict(default=None),
+        no_security_groups=dict(default=None),
+        allowed_address_pairs=dict(default=None),
+        extra_dhcp_opt=dict(default=None),
+        device_owner=dict(default=None),
+        device_id=dict(default=None),
+        state=dict(default='present', choices=['absent', 'present']),
+    )
+
+    module_kwargs = openstack_module_kwargs(
+        mutually_exclusive=[
+            ['no_security_groups', 'security_groups'],
+        ]
+    )
+
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
+                           **module_kwargs)
+
+    if not HAS_SHADE:
+        module.fail_json(msg='shade is required for this module')
+    name = module.params['name']
+    state = module.params['state']
+
+    try:
+        cloud = shade.openstack_cloud(**module.params)
+        port = cloud.get_port(name)
+
+        if module.check_mode:
+            module.exit_json(changed=_system_state_change(module, port))
+
+        changed = False
+        if state == 'present':
+            if not port:
+                network = module.params['network']
+                if not network:
+                    module.fail_json(
+                        msg="Parameter 'network' is required in Port Create"
+                    )
+                port_kwargs = _compose_port_args(module)
+                network_id = cloud.get_network(network)['id']
```

Comment:
get_network() can return None if no matching network is found. You probably should call module.fail_json() in that case.